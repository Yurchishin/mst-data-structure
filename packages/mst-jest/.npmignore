__test__
src
.babelrc
.eslintignore
.eslintrc
coverage
.gitignore
.npmignore
jest.config.js
rollup.config.js