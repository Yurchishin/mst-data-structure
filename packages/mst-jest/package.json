{
  "name": "@mst-ds/mst-jest",
  "version": "0.0.7",
  "description": "> TODO: description",
  "author": "Yurchishin",
  "homepage": "https://github.com/Yurchishin/mst-data-structure",
  "license": "MIT",
  "main": "lib/mst-jest.js",
  "umd:main": "lib/mst-jest.umd.js",
  "module": "lib/mst-jest.module.js",
  "browser": {
    "./lib/mst-jest.js": "./lib/mst-jest.js",
    "./lib/mst-jest.module.js": "./lib/mst-jest.module.js"
  },
  "unpkg": "lib/mst-jest.umd.min.js",
  "jsnext:main": "lib/mst-jest.module.js",
  "react-native": "lib/mst-jest.module.js",
  "directories": {
    "lib": "lib"
  },
  "files": [
    "lib/"
  ],
  "scripts": {
    "crean": "rm -rf lib/",
    "build": "rollup -c",
    "lint": "eslint --ext .ts src",
    "lint:fix": "eslint --fix --ext .js src"
  },
  "devDependencies": {
    "eslint": "^7.10.0",
    "mobx": "^5.15.4",
    "mobx-state-tree": "3.17.2",
    "rollup": "^2.21.0",
    "rollup-plugin-commonjs": "^10.1.0",
    "rollup-plugin-filesize": "^9.0.2",
    "rollup-plugin-node-resolve": "^5.2.0",
    "rollup-plugin-replace": "^2.2.0",
    "rollup-plugin-terser": "^6.1.0"
  },
  "peerDependencies": {
    "mobx": ">=4.8.0 <5.0.0 || >=5.8.0 <6.0.0",
    "mobx-state-tree": "3.17.2"
  },
  "keywords": [
    "mobx",
    "mobx-state-tree",
    "reactive",
    "state management",
    "data structure"
  ],
  "gitHead": "38957bca5fdad6f33b29dc295ae6d60a4063cc69",
  "publishConfig": {
    "access": "public"
  }
}
