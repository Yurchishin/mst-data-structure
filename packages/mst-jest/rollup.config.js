import { baseConfig } from "../../rollup.base-config"

const config = (outFile, format, mode) =>
    baseConfig({
        outFile,
        format,
        mode,

        input: "./src/index.js",
        globals: {
            mobx: "mobx",
            "mobx-state-tree": "mobxStateTree",
        },
        umdName: "mstJest",
        external: ["mobx", "mobx-state-tree"],
    })

export default [
    config("mst-jest.js", "cjs", "development"),
    config("mst-jest.min.js", "cjs", "production"),

    config("mst-jest.umd.js", "umd", "development"),
    config("mst-jest.umd.min.js", "umd", "production"),

    config("mst-jest.module.js", "es", "development"),
]