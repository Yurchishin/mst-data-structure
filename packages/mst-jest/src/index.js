import * as mobxStateTree from 'mobx-state-tree';
import {onPatch, onSnapshot, getSnapshot } from 'mobx-state-tree';

export {
    protect,
    unprotect,
    isProtected,
} from 'mobx-state-tree';

export const createMSTSnapshot = (model, ...options) => {
    expect(getSnapshot(model)).toMatchSnapshot(...options);
};

export const getEnvMock = (value = {}) => {
    jest.spyOn(mobxStateTree, 'getEnv').mockReturnValue(value);
};

export const getRootMock = (value = {}) => {
    jest.spyOn(mobxStateTree, 'getRoot').mockReturnValue(value);
};

export const mockMiddleware = (call, next) => next(call);

export const onPatchesAggregation = (model, ...options) => {
    const patches = [];

    const disposer = onPatch(model, patch => {
        patches.push(patch);
    });

    return () => {
        expect(patches).toMatchSnapshot(...options);
        disposer();
    };
};

export const onSnapshotAggregation = (model, ...options) => {
    const snapshots = [];

    const disposer = onSnapshot(model, patch => {
        snapshots.push(patch);
    });

    return () => {
        expect(snapshots).toMatchSnapshot(...options);
        disposer();
    };
};
