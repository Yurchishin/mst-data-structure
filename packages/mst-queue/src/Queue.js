import { types } from 'mobx-state-tree';
import LinkedList from '@mst/mst-linked-list';

const Queue = (name = '', Type, options = {}) => {
    const QueueModel = types
        .model(`${name}Queue`, {
            linkedList: LinkedList(`${name}Queue`, Type, options),
        })
        .views(self => ({
            get values() {
                return self.linkedList.values
            },
            get entries() {
                return self.linkedList.entries
            },
            get array() {
                return self.linkedList.array
            },
            get isEmpty() {
                return self.linkedList.isEmpty
            },
            toString(callback) {
                return self.linkedList.toString(callback)
            },
            get peek() {
                return !self.isEmpty ? self.linkedList.head : null
            },
            get peekPair() {
                return self.peek ? [self.peek.id, self.peek.value] : null
            },
            get peekValue() {
                return self.peekPair ? self.peekPair[1] : null
            },
            get peekId() {
                return self.peekPair ? self.peekPair[0] : null
            },
        }))
        .actions(self =>  {
            const enqueue = (id, value) => {
                self.linkedList.append(id, value);
                return self
            }

            const dequeue = () => {
                self.linkedList.deleteHead();
                return self
            }

            return {
                enqueue,
                dequeue,
            }
        })

    return types.optional(QueueModel, {})
}

export default Queue;