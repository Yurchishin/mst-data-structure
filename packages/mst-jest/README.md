## `mst-jest`

#### API

##### `createMSTSnapshot(node)` - created jest snapshot from model
##### `getEnvMock({...})` - mocked for `getEnv`'
##### `getRootMock({...})` - mocked for `getRoot`'
##### `onPatchesAggregation(node)` - aggregate all model patches, and take a jest snapshot'
##### `onSnapshotAggregation(node)` - aggregate all model snapshots, and take a jest snapshot'

#### Usage
```js
import { types } from "mobx-state-tree"
import { createMSTSnapshot, onPatchesAggregation, onSnapshotAggregation } from "@mst-ds/mst-jest"

describe('Tests', () => {
    it('snapshot test', () => {
        const rootStore = types.model({
            ...,
        })
        .actions(() => {...});

        createMSTSnapshot(list) // create jest snapshot
    });
it('snapshots test', () => {
        const rootStore = types.model({
            ...,
        })
        .actions(() => {...});

        const disposer = onPatchesAggregation(list)

        rootStore.action1();
        rootStore.action2();
        rootStore.action3();

        disposer() // create jest snapshot
    });
    it('patches test', () => {
        const rootStore = types.model({
            ...,
        })
        .actions(() => {...});

        const disposer = onSnapshotAggregation(list)

        rootStore.action1();
        rootStore.action2();
        rootStore.action3();

        disposer() // create jest snapshot
    });
});
```