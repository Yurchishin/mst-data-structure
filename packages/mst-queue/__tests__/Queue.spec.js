import { types } from 'mobx-state-tree';
import { onSnapshotAggregation, createMSTSnapshot } from '@mst-ds/mst-jest';
import Queue from '../src/index';

describe('Queue', () => {
    describe('model', () => {
        it('create(int)', () => {
            const list = Queue('Int', types.integer).create();

            createMSTSnapshot(list)
        });
        it('create in model(int)', () => {
            const rootStore = types.model('RootStore', {
                numbers: Queue('Int', types.integer),
            }).create()

            createMSTSnapshot(rootStore)
        });
        it('create in model(Todo)', () => {
            const Todo = types.model('Todo', {
                name: types.string,
                active: types.boolean,
            })

            const rootStore = types.model('RootStore', {
                todos: Queue('Todo', Todo),
            }).create()

            rootStore.todos
                .enqueue('3', {
                    name: 'name1',
                    active: false,
                })
                .enqueue('2', {
                    name: 'name2',
                    active: true,
                })
                .enqueue('1', {
                    name: 'name3',
                    active: false,
                })
                .dequeue()

            createMSTSnapshot(rootStore)
        });
    })
    describe('actions', () => {
        it('enqueue', () => {
            const list = Queue('Int', types.integer).create();

            const disposer = onSnapshotAggregation(list)

            expect(list.isEmpty).toBe(true);

            list.enqueue('1', 1);
            expect(list.peekId).toBe('1');

            list.enqueue('2', 2)
            expect(list.peekId).toBe('1');

            list.enqueue('3', 3)
            expect(list.peekId).toBe('1');

            disposer()
        });
        it('dequeue', () => {
            const list = Queue('Int', types.integer).create();

            const disposer = onSnapshotAggregation(list)

            expect(list.isEmpty).toBe(true);

            list.enqueue('1', 1)
                .enqueue('2', 2)
                .enqueue('3', 3)
                .dequeue()

            expect(list.peekId).toBe('2');

            disposer()
        });
    });
});
