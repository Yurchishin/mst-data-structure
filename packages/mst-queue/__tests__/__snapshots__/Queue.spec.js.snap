// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Queue actions dequeue 1`] = `
Array [
  Object {
    "linkedList": Object {
      "head": "1",
      "nodes": Object {
        "1": Object {
          "id": "1",
          "next": null,
          "value": 1,
        },
      },
      "tail": "1",
    },
  },
  Object {
    "linkedList": Object {
      "head": "1",
      "nodes": Object {
        "1": Object {
          "id": "1",
          "next": "2",
          "value": 1,
        },
        "2": Object {
          "id": "2",
          "next": null,
          "value": 2,
        },
      },
      "tail": "2",
    },
  },
  Object {
    "linkedList": Object {
      "head": "1",
      "nodes": Object {
        "1": Object {
          "id": "1",
          "next": "2",
          "value": 1,
        },
        "2": Object {
          "id": "2",
          "next": "3",
          "value": 2,
        },
        "3": Object {
          "id": "3",
          "next": null,
          "value": 3,
        },
      },
      "tail": "3",
    },
  },
  Object {
    "linkedList": Object {
      "head": "2",
      "nodes": Object {
        "2": Object {
          "id": "2",
          "next": "3",
          "value": 2,
        },
        "3": Object {
          "id": "3",
          "next": null,
          "value": 3,
        },
      },
      "tail": "3",
    },
  },
]
`;

exports[`Queue actions enqueue 1`] = `
Array [
  Object {
    "linkedList": Object {
      "head": "1",
      "nodes": Object {
        "1": Object {
          "id": "1",
          "next": null,
          "value": 1,
        },
      },
      "tail": "1",
    },
  },
  Object {
    "linkedList": Object {
      "head": "1",
      "nodes": Object {
        "1": Object {
          "id": "1",
          "next": "2",
          "value": 1,
        },
        "2": Object {
          "id": "2",
          "next": null,
          "value": 2,
        },
      },
      "tail": "2",
    },
  },
  Object {
    "linkedList": Object {
      "head": "1",
      "nodes": Object {
        "1": Object {
          "id": "1",
          "next": "2",
          "value": 1,
        },
        "2": Object {
          "id": "2",
          "next": "3",
          "value": 2,
        },
        "3": Object {
          "id": "3",
          "next": null,
          "value": 3,
        },
      },
      "tail": "3",
    },
  },
]
`;

exports[`Queue model create in model(Todo) 1`] = `
Object {
  "todos": Object {
    "linkedList": Object {
      "head": "2",
      "nodes": Object {
        "1": Object {
          "id": "1",
          "next": null,
          "value": Object {
            "active": false,
            "name": "name3",
          },
        },
        "2": Object {
          "id": "2",
          "next": "1",
          "value": Object {
            "active": true,
            "name": "name2",
          },
        },
      },
      "tail": "1",
    },
  },
}
`;

exports[`Queue model create in model(int) 1`] = `
Object {
  "numbers": Object {
    "linkedList": Object {
      "head": null,
      "nodes": Object {},
      "tail": null,
    },
  },
}
`;

exports[`Queue model create(int) 1`] = `
Object {
  "linkedList": Object {
    "head": null,
    "nodes": Object {},
    "tail": null,
  },
}
`;
